packages/protocol/contracts-0.8/common/EpochManager.sol

```diff
@@ -0,0 +1,338 @@
+  }
+
+  /// returns epoch processing state
+  function getEpochProcessingState()
+    external
+    view
+    returns (uint256, uint256, uint256, uint256, uint256)
+  {
+    return (
+      uint256(epochProcessing.status),
+      epochProcessing.perValidatorReward,
+      epochProcessing.totalRewardsVoter,
+      epochProcessing.totalRewardsCommunity,
+      epochProcessing.totalRewardsCarbonFund
+    );
+  }
+
+  function getElected() external view returns (address[] memory) {
+    return elected;
+  }
+
+  function getFirstBlockAtEpoch(uint256 epoch) external view returns (uint256) {
+    require(epoch >= firstKnownEpoch, "Epoch not known");
+    require(epoch <= currentEpochNumber, "Epoch not created yet");
+    return epochs[epoch].firstBlock;
+  }
+
+  function getLastBlockAtEpoch(uint256 epoch) external view returns (uint256) {
+    require(epoch >= firstKnownEpoch, "Epoch not known");
+    require(epoch < currentEpochNumber, "Epoch not finished yet");
+    return epochs[epoch].lastBlock;
+  }
+
+  function isBlocked() external view returns (bool) {
+    return isOnEpochProcess();
+  }
+
+  /**
+   * @notice Returns the storage, major, minor, and patch version of the contract.
+   * @return Storage version of the contract.
+   * @return Major version of the contract.
+   * @return Minor version of the contract.
+   * @return Patch version of the contract.
+   */
+  function getVersionNumber() external pure returns (uint256, uint256, uint256, uint256) {
+    return (1, 1, 0, 0);
+  }
+
+  /**
+   * @notice Sets the time duration of an epoch.
+   * @param newEpochDuration The duration of an epoch in seconds.
+   * @dev Can only be set by owner.
+   */
+  function setEpochDuration(uint256 newEpochDuration) public onlyOwner {
+    epochDuration = newEpochDuration;
+  }
+
+  function isTimeForNextEpoch() public view returns (bool) {
+    return block.timestamp >= epochs[currentEpochNumber].startTimestamp + epochDuration;
+  }
+
+  function isOnEpochProcess() public view returns (bool) {
+    return epochProcessing.status == EpochProcessStatus.Started;
+  }
+
+  function systemAlreadyInitialized() public view returns (bool) {
+    return initialized && elected.length > 0;
+  }
+
+  function allocateValidatorsRewards() internal {
+    uint256 totalRewards = 0;
+    IScoreReader scoreReader = getScoreReader();
+    IValidators validators = getValidators();
+
+    for (uint i = 0; i < elected.length; i++) {
+      uint256 validatorScore = scoreReader.getValidatorScore(elected[i]);
+      uint256 validatorReward = validators.computeEpochReward(
+        elected[i],
+        validatorScore,
+        epochProcessing.perValidatorReward
+      );
+      validatorPendingPayments[elected[i]] += validatorReward;
+      totalRewards += validatorReward;
+    }
+    // Mint all cUSD required for payment and the corresponding CELO
+    validators.mintStableToEpochManager(totalRewards);
+    // this should have a setter for the oracle.
```

Comment:
?

Reply:
this is an early contract that got overlooked, I meant to decouple SortedOracles in this contract. So instead of doing

`IOracle(address(getSortedOracles()))` it should be doing `getcUSDOracle()`

```
function getcUSDOracle() {
   return oracleAddress;
}
```

and `oracleAddress` should be set in the initializer (and then pointed to SortedOracles during the deployment).

Reply:
the address should probably be set in the `initializer` function